lib/vnet/ssh_handler.go

```diff
@@ -61,13 +81,93 @@ func (h *sshHandler) handleTCPConnectorWithTargetConn(
 	connector func() (net.Conn, error),
 	targetConn net.Conn,
 ) error {
-	// For now we accept the incoming TCP conn to indicate that the node exists,
-	// but SSH connection forwarding is not implemented yet so we immediately
-	// close it.
+	hostCert, err := h.getHostCert(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
 	localConn, err := connector()
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	localConn.Close()
+	defer localConn.Close()
+
+	// For now we accept the incoming SSH connection but forwarding to the
+	// target is not implemented yet so we immediately close it.
+	serverConfig := &ssh.ServerConfig{
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			if !sshutils.KeysEqual(h.cfg.sshProvider.trustedUserPublicKey, key) {
+				return nil, trace.AccessDenied("SSH client public key is not trusted")
+			}
+			return &ssh.Permissions{}, nil
+		},
+	}
+	serverConfig.AddHostKey(hostCert)
```

Comment:
We could add the underlying CA key here too, to also support (in a way that's more prone to errors) clients that only understand keys and not certificates. WDYT? We'd still only list the key as `@cert-authority` in `vnet_known_hosts`, which would mean that by default the key by itself is going to be ignored.

(if ssh ends up accepting the raw key for any host if it's marked `@cert-authority` we could just use a different host key)

Reply:
what's the point if we don't add it to `vnet_known_hosts`?